src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,188 @@
+		const [response, body] = await requestAsync({
+			url: `http://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/query_range`,
+			headers: {
+				'X-Scope-OrgID': ctx.belongs_to__application,
+			},
+			method: 'GET',
+			qs: {
+				query: this.getDeviceQuery(ctx),
+				limit: _.isFinite(count) ? count : 1000,
+			},
+			json: true,
+		});
+		if (response.statusCode === 200) {
+			return this.fromLokiStreamsToDeviceLogs(body.data.result);
+		} else {
+			throw new BadRequestError('Query failed');
+		}
+	}
+
+	public async publish(ctx: LogWriteContext, logs: DeviceLog[]): Promise<any> {
+		return requestAsync({
+			url: `http://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/push`,
+			headers: {
+				'X-Scope-OrgID': ctx.belongs_to__application,
+			},
+			method: 'POST',
+			json: {
+				streams: logs.map((log) => this.toLokiStream(ctx, log), this),
+			},
+		});
+	}
+
+	public subscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		if (!this.webSockets.has(key)) {
+			const tailUri =
+				`ws://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/tail` +
+				`?query=${encodeURIComponent(this.getDeviceQuery(ctx))}&start=${String(
+					Date.now(),
+				)}000000`;
+			const ws = new WebSocket(tailUri, {
+				headers: { 'X-Scope-OrgID': ctx.belongs_to__application },
+			});
+			ws.on('message', (data: string) => {
+				this.fromLokiStreamsToDeviceLogs(JSON.parse(data).streams).map(
+					(log: DeviceLog) => {
+						this.subscriptions.emit(key, log);
+					},
+				);
+			});
+			ws.on('close', () => {
+				this.webSockets.delete(key);
+			});
+			this.webSockets.set(key, ws);
+		}
+		this.subscriptions.on(key, subscription);
+	}
+
+	public unsubscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		this.subscriptions.removeListener(key, subscription);
+		if (!this.webSockets.has(key)) {
+			const ws = this.webSockets.get(key);
+			if (ws) ws.close();
+		}
+	}
+
+	private getDeviceQuery(ctx: LogContext) {
+		return `{device_id="${ctx.id}"}`;
+	}
+
+	private getKey(ctx: LogContext, suffix = 'logs') {
+		return `device:${ctx.id}:${suffix}`;
+	}
+
+	private fromLokiStreamsToDeviceLogs(streams: LokiStream[]): DeviceLog[] {
+		return _.concat(
+			[],
+			...streams.map((stream: LokiStream) => this.fromLokiStream(stream)),
+		);
+	}
+
+	private fromLokiStream(stream: LokiStream): DeviceLog[] {
+		try {
+			return stream.values.map((value) => {
+				const [, message] = value;
+				const log = JSON.parse(message);
+				if (log.version !== VERSION) {
+					throw new Error(
+						`Invalid Redis serialization version: ${JSON.stringify(log)}`,
+					);
+				}
+				if (log.serviceId === null) {
+					delete log.serviceId;
+				}
+				delete log.version;
+				return log as DeviceLog;
+			});
+		} catch (err) {
+			captureException(err, `Failed to deserialize a Loki log: ${stream}`);
+			return [];
+		}
+	}
+
+	private toLokiStream(ctx: LogContext, log: DeviceLog): LokiStream {
+		try {
+			skhema.validate(schema, log);
```

Comment:
I think skhema has performance issues and this could potentially bring the api down - I'd want the performance to be tested/validated before merging rather than waiting for production

Reply:
I will re-implement this using `io-ts` instead.

Reply:
I thought `io-ts` might be slow aswell. So instead I removed `skema` and did not add any validation replacement. Do you think validation is necessary here?

Reply:
We use avro for the redis logs and so I'd be happy to use that without additional performance testing since it's already validated to be fast enough, otherwise I'm also happy with other validation I just want it to have performance tested as there's a lot of logs coming into the system. And yes I think it makes sense to have validation of the logs rather than potentially letting garbage in which could cause issues of its own

Reply:
Updated with a custom function for basic validation and speed